package.json

```diff
@@ -252,7 +256,8 @@
 		],
 		"packages/**/*.js": [
 			"node ./bin/api-docs/update-readmes.js",
-			"node ./bin/api-docs/are-readmes-unstaged.js"
+			"node ./bin/api-docs/are-readmes-unstaged.js",
+			"node ./bin/packages/lint-staged-typecheck.js"
```

Comment:
901ce5d07a01dc8a4691577f3f87e0ff796214d5 removed the custom build script and relies on a root `tsconfig.json` that references each project. This simplifies the type build significantly.

I'm inclined to have the `lint-staged` execute `npm run packages:build-types`. That would remove an additional custom script introduced here.

Once the TypeScript cache is warm, incremental builds are quick or instant. Running the entire type build has the benefit of detecting when type incompatibilities across packages are introduced, while the current version only detects that the _changed package itself_ typechecks and issues between packages would surface later on CI when the types are compiled.

@aduth Curious to hear your thoughts.

Reply:
I attempted to do this and it's quite difficult because `lint-staged` always passes the changed files at the end, meaning we wind up with a command like the following:

```sh
npm run build:package-types "…/gutenberg/packages/a11y/src/index.js"

# tsc --build --verbose "…/gutenberg/packages/a11y/src/index.js"
# error TS6053: File '…/gutenberg/packages/a11y/src/index.js/tsconfig.json' not found.
```

If `lint-staged` is defined as a `js` file, [lint-staged tasks can be declared as functions](https://github.com/okonet/lint-staged#using-js-functions-to-customize-tasks) which makes it trivial to ignore the files input.

Are we open to that? I'd prefer to handle that in its own PR.

Reply:
> I'm inclined to have the `lint-staged` execute `npm run packages:build-types`. That would remove an additional custom script introduced here. [...] Once the TypeScript cache is warm, incremental builds are quick or instant.

Is the motivation to avoid maintaining the custom script, or because it may be faster to run a (warmed) full build than to typecheck individual package/files?

I'm motivated to keep these pre-commits fast. I'm not overly concerned with the maintenance burden of a custom script, if it means we can have faster/more granular checks.

>Running the entire type build has the benefit of detecting when type incompatibilities across packages are introduced, while the current version only detects that the changed package itself typechecks and issues between packages would surface later on CI when the types are compiled.

I agree it would be good to have more immediate feedback about potential issues. If it can be fast enough to not become a burden, then I like the idea. I'd also be content if at least this is considered in the full Travis build.

Aside: Slow pre-commit has become a pet-peeve of mine lately (see #18840), in case that helps contextualize my thoughts in approaching this.

Reply:
> Is the motivation to avoid maintaining the custom script

This is a motivation

> or because it may be faster to run a (warmed) full build than to typecheck individual package/files?

A warm build of everything will likely be almost identical to a warm typecheck of changed files. Only the necessary things would be rebuilt.

The difference between a hot and cold cache is huge. The difference between a single package and all packages with a cold cache is also very large:

```sh
# Cold cache all
npm run clean:packages; time npm run build:package-types
# real    0m37.884s
# user    0m46.316s
# sys     0m1.311s

# Cold cache only URL (random package)
npm run clean:packages; time npm run build:package-types -- packages/url
# real    0m5.567s
# user    0m8.644s
# sys     0m0.353s

# Hot cache all
npm run build:package-types; time npm run build:package-types
# real    0m0.515s
# user    0m0.356s
# sys     0m0.091s
```

If a package has changed, it will be rebuilt. If it's dependencies have not been built, they will be built first. Given that rebuilding an unchanged package with a warm cache is practically free, the difference between building a single package and building all packages should be insignificant.

I agree these staged checks should be near immediate, and I don't feel too strongly about this.

We might try to persist TypeScript caches longer. TypeScript does an excellent job of doing absolutely nothing very quickly if it doesn't need to. I suspect it's only important to clean these types files before publish, so perhaps `tsc --build --clean` should be removed from `build:packages` and into the `publish:…` scripts and we could.

Reply:
> We might try to persist TypeScript caches longer.

I've done this in 12302c0 where type cleaning is moved to publish scripts.

Reply:
51195f349f5ef936b93b7c74de3addde0d623c8b proposes typechecking all modules on lint-staged. Feel free to play with it, but if you feel strongly we can roll this back. I don't feel strongly one way or the other.